<!DOCTYPE html>
<html>
<head>
<title>scrnaseq_processing_workflows.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="workflow-name">[workflow name]</h1>
<ul>
<li><a href="#description">Description</a></li>
<li><a href="#diagram">Diagram</a></li>
<li><a href="#user-guide">User guide</a></li>
<li><a href="#background-and-tutorials">Background and Tutorials</a></li>
<li><a href="#licences">Licence(s)</a></li>
<li><a href="#acknowledgementscitationscredits">Acknowledgements/citations/credits</a></li>
</ul>
<hr>
<h1 id="description">Description</h1>
<p>This document describes how to use some scanpy-based scRNAseq workflows on galaxy Australia.</p>
<p>The aim of these workflows is to handle the routine ‘boring’ part of single cell RNAseq data processing. It will produces an ‘AnnData’ object, which can then be used as a base for downstream analysis – either within galaxy or outside of it. AnnData is a standard format used by the ‘scanpy’ python package.</p>
<p>These workflows represent just one way of processing data for a ‘typical’ scRNAseq experiment – there are many other options!</p>
<p>This document describes 3 sub-workflows for processing single cell RNAseq data with scanpy</p>
<ul>
<li><strong>Load counts matrix:</strong> <a href="https://usegalaxy.org.au/u/s.williams/w/scrnaseq-load-counts-matrix-subworkflow">link</a>   This workflow adds a sample name, which enables multi-sample analyses</li>
<li><strong>Single cell QC:</strong> <a href="https://usegalaxy.org.au/u/s.williams/w/scrnaseqcellqc">link</a>  This workflow generates some basic QC plots and applies filtering</li>
<li><strong>Single cell QC to basic processing:</strong> <a href="https://usegalaxy.org.au/u/s.williams/w/copy-of-scanpyqcplusdraft">link</a> This generates a UMAP, does clustering and calculates cluster marker genes.</li>
</ul>
<p>For single sample experiments, there is a streamlined workflow that runs all 3 sub-workflows all at once</p>
<ul>
<li><strong>Single sample workflow:</strong> <a href="https://usegalaxy.org.au/u/s.williams/w/copy-of-scrnaseqcountsmatrixtoqc">link</a> This workflow loads counts matrix, does some basic processing, suitable for a single sample.</li>
</ul>
<p>These workflows are all available on galaxy australia.</p>
<hr>
<h1 id="analysis-overview">Analysis overview</h1>
<p><img src="/images/workflow_diagram_simple.png" alt="Processing flowchart"></p>
<ol>
<li>Start with fastq files</li>
<li>Cellranger|starSOLO will align the reads to the genome, and make a table of the number of times each gene is counted per cell.</li>
<li>Perform some basic QC on the counts matrix, and filter out ‘cells’ that have too little RNA counts or too much mitochondrial gene content</li>
<li>Run some basic single cell analyses: Normalisation, PCA, UMAP, clustering and identify cluster markesr</li>
<li>The resulting AnnData object can be analysed further. See ‘Next steps’</li>
</ol>
<hr>
<h1 id="example-output">Example output</h1>
<p>When run in full, these workflows produce the following main outputs</p>
<ul>
<li>A processed <a href="https://anndata.readthedocs.io/en/latest/">AnnData</a> file, which contains gene expression and annotation information ready for downstream analysis.</li>
<li>Tables of 'marker' gene information - to aid detemination of cell types present in the experiment</li>
<li>Two report summaries.
<ul>
<li>Single cell QC report <a href="https://usegalaxy.org.au/u/s.williams/p/invocation-report-ede7b160ea86b66e">(example)</a> : This shows QC metrics at the cell level, to evaluate filtering thresholds and data quality.</li>
<li>Single cell basic processing report <a href="https://usegalaxy.org.au/u/s.williams/p/invocation-report-21aa7559fbcd167e">(example)</a> : Some basic UMAP, clustering and cluster marker results to begin an analysis with.</li>
</ul>
</li>
</ul>
<hr>
<h1 id="user-guide">User guide</h1>
<h2 id="running-a-single-sample-workflow">Running a single sample workflow</h2>
<p>When there is only a single biological sample in a study, there is a streamlined workflow.</p>
<ul>
<li><a href="https://usegalaxy.org.au/u/s.williams/w/copy-of-scrnaseqcountsmatrixtoqc">Single sample workflow</a></li>
</ul>
<ol>
<li>Input <em>will be</em> fastq files.</li>
<li>Search for the <em>scRNAseq Single sample workflow</em> under 'workflow' and run it. You'll be prompted to customise any filtering parameters, and choose a sensible name for the biological sample.</li>
</ol>
<p><img src="./images/screen_single_sample_launch.png" alt="Single Sample Launch prompt"></p>
<ol start="3">
<li>
<p>This pipeline should take a [few minutes/few hours] to run.</p>
</li>
<li>
<p>Return to galaxy, and look up the run, or invocation, of this workflow:</p>
</li>
</ol>
<p><em>User Menu &gt; Workflow Invocations</em></p>
<p>This brings up the history of workflow invocations. This particular workflow runs a number of sub-workflows. The <strong>Cell QC</strong> and <strong>QC to Basic Processing</strong> subworkflows produce reports, which can be viewed shared and saved.</p>
<ul>
<li>Cell QC : Generates the cell level QC plots</li>
<li>QC to Basic Processing: Shows the subsetquent UMAP, clustering and marker information. Includes links to download the processed AnnData object for downstream work</li>
</ul>
<p><img src="./images/screen_invocations.png" alt="workflow invocations screen"></p>
<h2 id="running-a-multi-sample-workflow">Running a multi-sample workflow</h2>
<p>With multi-sample experiments, each sample is loaded independently and then combined. The overall method is the same, but the QC and processing steps are run separately.</p>
<p><img src="./images/workflow_diagram_large.jpg" alt="subworkflows"></p>
<ol>
<li>Upload the raw data for one sample. Then run The ‘scRNAseq: Load counts matrix’ workflow – this will prompt you for a sample name that will be used throughout.</li>
</ol>
<p><img src="./images/screen_load_counts_matrix_launch.png" alt="Load counts matrix launch"></p>
<pre class="hljs"><code><div>**IF USING the 3 file input, describe how to change barcodes.tsv to tabular, and note that features == genes, and that .tsv.gz will work fine. ** 
</div></code></pre>
<p>This part of the workflow will load the counts matrix into an anndata object, and then adds an extra column in the metadata called ‘sample’. This means the sample information can be tracked when multiple samples are combined.</p>
<p>The AnnData object that it produces in your history will probably be named something like ‘Manipulate AnnData (add_annotation) on data 20 and data 17’. You may choose to rename this object via the ‘edit attributes’ option in the history panel, so its easier to find later.</p>
<p><img src="./images/screen_rename.png" alt="rename"></p>
<ol start="2">
<li>
<p>In the same history, repeat for all other samples.</p>
</li>
<li>
<p>Next, join all samples with the ‘Manipulate AnnData object’ Tool (search on the tools pane on the left).</p>
</li>
</ol>
<p>This tool can do several different operations – listed under ‘Function to manipulate the object’, but we want the default; “Concatenate along the observation axis”. This combines cells (observations) from multiple sample runs.</p>
<p>Choose the AnnData object of one of your samples in the  ‘Annotated data matrix’ dropdown. Then, choose the rest of your samples under ‘Annotated matrix to add’. Use ctrl-select / option-select to highlight multiple samples.</p>
<p><em>Note:</em> Be careful not to select the sample in the ‘Annotated data matrix’ dropdown again – else it will be joined to itself! In this example there is only two samples, so only pbmc8k is selected to be added to pbmc1k.</p>
<p><img src="./images/screen_multi_merge.png" alt="join anndatas"></p>
<p>A combined anndata object will created in your history.</p>
<ol start="4">
<li>Next, run the <strong>scRNAseq Cell QC</strong> workflow on your combined anndata object.</li>
</ol>
<p>This workflow plots some basic cell-level QC thresholds, and applies the QC thresholds to produce a filtered Anndata object.</p>
<p><img src="./images/screen_cellQC_launch.png" alt=""></p>
<ol start="5">
<li>Once it finishes running, view the report (Go to User menu &gt; Invocations to find it).</li>
</ol>
<p>You’ll notice you can see each sample plotted separately in the QC plots. You may elect to rerun with tweaked thresholds (e.g. higher minimum counts threshold) once you've seen this output.</p>
<p><img src="./images/cell_qc_plot.png" alt=""></p>
<ol start="6">
<li>If you are happy with the filtering thresholds, you can  launch the next workflow, <strong>scRNAseq QC to Basic Processing</strong> to do some routine single cell calculations.  It only asks for the filtered Anndata Object (typically the last AnnData in your history, with a name ending in ‘Filtered Cells Anndata’.)</li>
</ol>
<p><img src="./images/screen_qc_to_basic_processing_launch.png" alt="launch basic processing"></p>
<ol start="7">
<li>This takes a few minutes to run. Once finished, return to the invocations page to see the QC to Basic Processing report, as per the single sample workflow.</li>
</ol>
<p>The first umap now shows the different samples that make up the data.</p>
<p><img src="./images/umap_by_sample.png" alt="umap samples"></p>
<h2 id="next-steps">Next steps</h2>
<p>The AnnData object generated is ready for analysis! Options include</p>
<ul>
<li>CellXgene : CellXgene is a tool for browsing and exploring single cell data. It use the AnnData object, and all of the annotation stored within it (expression, clusters, sample names). See
<a href="https://cellxgene.cziscience.com/docs/01__CellxGene">CellXgene documentation</a>, and a <a href="https://icbi-lab.github.io/cellxgene-user-guide/">user guide</a>. This can be launched within galaxy.</li>
<li>Scanpy : <a href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a> is a python based suite of tool for working with single cell data.
<ul>
<li>Many scanpy functions are available within galaxy (and are used within this workflow). Explore the <a href="https://training.galaxyproject.org/training-material/topics/single-cell/">galaxy scRNAseq tutorials</a> for more information</li>
<li>Or, you download your AnnData object and work with the scanpy toolkit directly, using python on your computer or other server. This can provide more fine-grained control of your analysis. There are <a href="https://scanpy.readthedocs.io/en/stable/tutorials.html">many tutorials</a> to work from.</li>
</ul>
</li>
</ul>
<p>Note that there are toolkits other than scanpy (e.g. Seurat, SingleCellExperiment objects) which may not be directly compatible without conversions.</p>
<hr>
<h1 id="background-and-tutorials">Background and Tutorials</h1>
<p>For more general information about <strong>single cell RNAseq processing on galaxy</strong>; there are some excellent tutorials to be found here on the <a href="https://training.galaxyproject.org/training-material/topics/single-cell/">galaxy training website scRNA section</a>. The workflow implemented here is heavily influenced by the <a href="https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-scanpy-pbmc3k/tutorial.html">Clustering 3kPBMCs with Scanpy tutorial</a></p>
<p>More general information on <strong>using galaxy</strong> can be found on the <a href="https://training.galaxyproject.org/training-material/">galaxy training website</a></p>
<p>There are many general resources online about the princials of single cell analysis. The <a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">Scanpy preprocessing and clustering tutorial</a> may be of particular use becuase it describes the scanpy methods used in this workflow. Even if you don't use the python code, it works through and explains many of the plots these workflows generate.</p>
<hr>
<h1 id="licenses">License(s)</h1>
<hr>
<h1 id="acknowledgementscitationscredits">Acknowledgements/citations/credits</h1>
<p>The workflow implemented here is heavily influenced by the <a href="https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-scanpy-pbmc3k/tutorial.html">Clustering 3kPBMCs with Scanpy tutorial</a></p>
<hr>

</body>
</html>
